src/shogun/lib/parameter_observers/ObservedValue.h

```diff
@@ -0,0 +1,209 @@
+
+/**
+ * Definitions of basic object with are needed by the Parameter
+ * Observer architecture.
+ */
+namespace shogun
+{
+	/* Timepoint */
+	typedef std::chrono::steady_clock::time_point time_point;
+
+	/* Type of the observed value */
+	enum SG_OBS_VALUE_TYPE
+	{
+		TENSORBOARD
+	};
+
+	/**
+	 * Observed value which are emitted by algorithm.
+	 *
+	 * This class can be used to implement custom observed values.
+	 * For instance, one can code an observed value which can be
+	 * used to retrieve data from CrossValidation, CMachines etc.
+	 */
+	class ObservedValue
+	{
+	public:
+		ObservedValue(){};
+
+		/**
+		 * Constructor
+		 * @param step step
+		 * @param name param's name
+		 * @param value Any-wrapped value of the param
+		 */
+		ObservedValue(
+		    int64_t step, std::string& name, Any value, SG_OBS_VALUE_TYPE type)
+		    : m_step(step), m_name(name), m_value(value), m_type(type)
+		{
+		}
+
+		/**
+		 * Constructor for observed value which don't need a name or step
+		 * @param value the value of this observed parameters
+		 * @param type the type of this observed parameters
+		 */
+		ObservedValue(Any value, SG_OBS_VALUE_TYPE type)
+		    : m_step(0), m_name(""), m_value(value), m_type(type)
+		{
+		}
+
+		~ObservedValue(){};
+
+		/**
+		 * Get the step
+		 * @return an integer representing the step
+		 */
+		int64_t get_step() const
+		{
+			return m_step;
+		}
+
+		/**
+		 * Set the step
+		 * @param step step
+		 */
+		void set_step(int64_t step)
+		{
+			m_step = step;
+		}
+
+		/**
+		 * Get the param's name
+		 * @return param's name
+		 */
+		const std::string& get_name() const
+		{
+			return m_name;
+		}
+
+		/**
+		 * Set the param's name
+		 * @param name
+		 */
+		void set_name(const std::string& name)
+		{
+			m_name = name;
+		}
+
+		/**
+		 * Get the Any-wrapped value
+		 * @return Any-wrapped value
+		 */
+		const Any& get_value() const
+		{
+			return m_value;
+		}
+
+		/**
+		 * Set the param's value
+		 * @param value
+		 */
+		void set_value(const Any& value)
+		{
+			m_value = value;
+		}
+
+		/**
+		 * Get the type of this ObservedValue
+		 * @return observed value type
+		 */
+		SG_OBS_VALUE_TYPE get_type() const
+		{
+			return m_type;
+		}
+
+		/**
+		 * Set the observed value type
+		 * @param type type of this observed value
+		 */
+		void set_type(const SG_OBS_VALUE_TYPE type)
+		{
+			m_type = type;
+		}
+
+	protected:
+		/** ObservedValue step (used by Tensorboard to print graphs) */
+		int64_t m_step;
+		/** Parameter's name */
+		std::string m_name;
+		/** Parameter's value */
+		Any m_value;
+		/** ObservedValue type */
+		SG_OBS_VALUE_TYPE m_type;
+	};
+
+	/**
+	 * Helper method to generate an ObservedValue (TensorBoard oriented)
+	 * @param step the step
+	 * @param name the param's name we are observing
+	 * @param value the param's value
+	 * @return an ObservedValue object initialized
+	 */
+	SG_FORCED_INLINE ObservedValue
+	make_observation(int64_t step, std::string& name, Any value)
```

Comment:
this could be a static function of ObservedValue, just like a factory function and returns a std::shared_ptr/some/*